feat: logger (#7238)
vuestorefront/vue-storefront: docs/content/3.middleware/2.guides/10.logging.md

  /**
   * Own implementation of logger to be used internally.
   *
   * @remarks If provided then other options won't be used.
   */
  handler?: LoggerInterface;
}
```

#### 1. level

This option sets the log level based on the [RFC 5424](https://datatracker.ietf.org/doc/html/rfc5424) syslog standard. It defines the severity of log messages that should be recorded. Available log levels are:

- **emergency**: The highest severity level, indicating a system-wide emergency.
- **alert**: Indicates an urgent condition that requires immediate attention.
- **critical**: Highlights critical issues that need immediate action.
- **error**: Used to log error messages that could affect functionality.
- **warning**: Logs warnings about potential issues that should be monitored.
- **notice**: Records significant but non-critical events.
- **info**: Logs general informational messages indicating normal operation.
- **debug**: The most detailed log level, useful for tracing and debugging the application.

**Example:**
Setting the log level to `error` ensures only error messages and above (i.e., critical, alert, emergency) will be captured:

```json
{
  "logger": {

>     "level": "error"

Reviewer: ```suggestion
    "level": "error"
```
one last catch, wouldn't `verbosity` be more descriptive as we're talking specifically about log verbosity levels here/

# Logging

The middleware application provides a logger instance that automatically attaches metadata related to the scope of each call. By adding this contextual data, the logger makes it significantly easier to trace the origin of logs or errors, simplifying the debugging and monitoring process across the application.

## Using logger

The middleware application provides access to the logger in various parts of the system, such as in extensions and integrations. This flexibility allows you to log important events, errors, and other data throughout the lifecycle of your application.

In this section, we will explore how to access and use the logger in different parts of the application.

> The below examples will demonstrate how to use the application logger in a way that ensures that correct metadata is conveyed in the logs. To achieve that we will use `getLogger` function that will extract the logger instance from the passed argument.

Reviewer 2: ```suggestion
The examples below demonstrate how to use the application logger to ensure that correct metadata is conveyed in the logs. To achieve that, we will use the `getLogger` function to extract the logger instance from the passed argument.
```

Author: fixed

:::tip
The logger is integrated into the middleware to provide additional metadata about the scope of call with every log out of the box. In case of error, it also provides an error boundary.
:::

### Basic Logger usage

To use the logger, you need to import the `getLogger` function from the middleware package. Depending on the context, you can obtain the logger from different sources, such as the `alokai` object, `context` object, or `params` object. All such cases are covered in the following examples. For this example, we will use the `context` object.

```ts
import { getLogger } from "@vue-storefront/middleware";

// Assume that `context` is available in the current scope
const logger = getLogger(context);
logger.info("Middleware is running!");
```

Prints:

```json
{
  "message": "Middleware is running!",
  "timestamp": "2024-10-22T19:04:18.791Z",
  "severity": "INFO",
  "alokai": {
    "context": "middleware",
    "scope": {

>       ...

Reviewer 2: ```suggestion
      // ...
```

    }
  }
}
```

The provided logger has a second optional argument that allows you to include custom metadata to be attached to the JSON output.
For example:

```ts
logger.info("I've just fetched a user!", {
  userId: "123",
});
```

Prints:

```json
{
  "message": "I've just fetched a user!",
  "timestamp": "2024-10-22T19:04:18.791Z",
  "severity": "INFO",
  "alokai": {
    "context": "middleware",
    "scope": {

>       ...

Reviewer 2: ```suggestion
      // ...
```

};
```

:::warning
Attempt of using `hooksLogger` inside `beforeCall` or other hooks via closure corrupts information about scope of call of the log. Always use logger provided in params to the hook.
:::

### extendApiMethods

Inside extended api methods, obtain logger from `context` using `getLogger` function.

```ts
import { getLogger } from "@vue-storefront/middleware";

const testingExtension = {
  name: "some-extension",
  extendApiMethods: {
    addedCustomEndpoint(context) {
      const logger = getLogger(context);
      logger.info("You can log inside newly created handler!");
      return {
        // ...
      };
    },
  },
};
```

:::tip
You can derive logger from context the same way if you are creating an integration.
:::

### onCreate

The hook got a new parameter from which you can derive the logger with help of `getLogger` function.

```ts
import { type AlokaiContainer, getLogger } from "@vue-storefront/middleware";

const onCreate = async (
  config: Record<string, unknown> = {},
  alokai: AlokaiContainer
) => {
  const logger = getLogger(alokai);
  logger.info("You can log inside onCreate when creating a new integration");

  return {
    // ...
  };
};
```

> ### init

Reviewer 2: ```suggestion
### Init
```

- **debug**: The most detailed log level, useful for tracing and debugging the application.

**Example:**
Setting the log verbosity level to `error` ensures only error messages and above (i.e., critical, alert, emergency) will be captured:

```json
{
  "logger": {
    "verbosity": "error"
  }
}
```

For a development environment, you might prefer a more verbose log level like `debug`:

```json
{
  "logger": {
    "verbosity": "debug"
  }
}
```

#### 2. includeStackTrace

This option specifies whether the stack trace should be included in the log messages. If set to `true`, stack traces will be included in error messages, which can be helpful during debugging. If set to `false`, logs will be more concise.

```json
{
  "logger": {
    "includeStackTrace": true
  }
}
```

### Global configuration

To configure the logger globally, add a `logger` object to the top-level configuration of the middleware. This configuration will apply to all integrations and parts of the application unless specifically overridden. Here’s an example of a global configuration where the stack trace is included in the logs, and the log level is set to `debug`:

> ```js[middleware.config.js]

Reviewer 2: Example should be wrapped with 
```
export const config = {
```

Author: ok, fixed

{
  "logger": {
    "includeStackTrace": true,
    "verbosity": "debug"
  },
  // The rest of the middleware configuration
  "integrations": {
    // ...
  }
}
```

This global logger configuration ensures that all logs, regardless of the integration or scope, will follow these settings.

### Per integration configuration

If you need different logging settings for specific integrations, you can configure the logger separately within each integration. The logger configuration should be added at the same level as `location` and `configuration` keys within the integration’s object.

For example, here is how you would configure the logger for an integration with the key `commerce`:

```js[middleware.config.js]

> {

Reviewer 2: Also here should be wrapped with 
```
export const config = {
```

Author: fixed
